pkg/porter/explain.go

```diff
@@ -0,0 +1,291 @@
+	"github.com/deislabs/porter/pkg/context"
+	"github.com/deislabs/porter/pkg/printer"
+	"github.com/olekukonko/tablewriter"
+	"github.com/pkg/errors"
+)
+
+type ExplainOpts struct {
+	BundleLifecycleOpts
+	printer.PrintOptions
+}
+
+// PrintableBundle holds a subset of pertinent values to be explained from a bundle.Bundle
+type PrintableBundle struct {
+	Name        string                         `json:"name" yaml:"name"`
+	Description string                         `json:"description,omitempty" yaml:"description,omitempty"`
+	Version     string                         `json:"version" yaml:"version"`
+	Parameters  map[string]PrintableParameter  `json:"parameters,omitempty" yaml:"parameters,omitempty"`
+	Credentials map[string]PrintableCredential `json:"credentials,omitempty" yaml:"credentials,omitempty"`
+	Outputs     map[string]PrintableOutput     `json:"outputs,omitempty" yaml:"outputs,omitempty"`
+	Actions     map[string]PrintableAction     `json:"customActions,omitempty" yaml:"customActions,omitempty"`
+}
+
+type PrintableCredential struct {
+	Description string `json:"description" yaml:"description"`
+	Required    bool   `json:"required" yaml:"required"`
+}
+
+type PrintableOutput struct {
+	Type        interface{} `json:"type" yaml:"type"`
+	ApplyTo     string      `json:"applyTo" yaml:"applyTo"`
+	Description string      `json:"description" yaml:"description"`
+}
+
+type PrintableParameter struct {
+	Type        interface{} `json:"type" yaml:"type"`
+	Default     interface{} `json:"default" yaml:"default"`
+	ApplyTo     string      `json:"applyTo" yaml:"applyTo"`
+	Description string      `json:"description" yaml:"description"`
+	Required    bool        `json:"required" yaml:"required"`
+}
+
+type PrintableAction struct {
+	Modifies bool `json:"modifies" yaml:"modifies"`
+	// Stateless indicates that the action is purely informational, that credentials are not required, and that the runtime should not keep track of its invocation
+	Stateless bool `json:"stateless" yaml:"stateless"`
+	// Description describes the action as a user-readable string
+	Description string `json:"description" yaml:"description"`
+}
+
+func (o *ExplainOpts) Validate(args []string, cxt *context.Context) error {
+	err := o.sharedOptions.Validate(args, cxt)
+	if err != nil {
+		return err
+	}
+	err = o.ParseFormat()
+	if err != nil {
+		return err
+	}
+	if o.Tag != "" {
+		o.File = ""
+		o.CNABFile = ""
+
+		return o.validateTag()
+	}
+	return nil
+}
+
+func (p *Porter) Explain(o ExplainOpts) error {
+	err := p.prepullBundleByTag(&o.BundleLifecycleOpts)
+	if err != nil {
+		return errors.Wrap(err, "unable to pull bundle before invoking credentials generate")
+	}
+
+	err = p.applyDefaultOptions(&o.sharedOptions)
+	if err != nil {
+		return err
+	}
+	err = p.ensureLocalBundleIsUpToDate(o.bundleFileOptions)
+	if err != nil {
+		return err
+	}
+	bundle, err := p.CNAB.LoadBundle(o.CNABFile, o.Insecure)
+	// Print Bundle Details
+
+	pb, err := generatePrintable(bundle)
```

Comment:
nit: I suggest splitting this function up into two. One that handles everything up to this point: pulling, handling the options, generating the dataset. Then another that handles printing it out. We have been using that pattern elsewhere, for `porter mixins list` is split into print mixins and list mixins:

https://github.com/deislabs/porter/blob/7da5cffdf3581c27dfab03af88df2df005baa02d/pkg/porter/mixins.go#L16

https://github.com/deislabs/porter/blob/7da5cffdf3581c27dfab03af88df2df005baa02d/pkg/porter/mixins.go#L42

I realize we aren't perfectly consistent. There are examples in the codebase of doing this differently, but I'm picking the one I think works best. This helps with testing (we aren't testing printing, just the data collected), and also ensures that we are making porter reusable for others to build upon.